File: C++/Array/Sorting/Mergesort.cpp
//Time complexity-> Best case - nlogn
//                  Average case - nlogn
//                  Worst case - nlogn

//It is a stable algorithm.
#include<iostream>
using namespace std;

void merge(int arr[], int l, int mid, int r){
	int n1 = mid - l +1;
	int n2 = r - mid;
	int a[n1];
	int b[n2];
	
	for(int i = 0; i < n1; i++){
		a[i] = arr[l + i];
	}
	
	for(int j = 0; j < n2; j++){
		b[j] = arr[mid + 1 + j];
	}
	
	int i = 0;
	int j = 0;
	int k = l;
	while(i < n1 && j < n2){
		if(a[i] < b[j]){
			arr[k] = a[i];
			k++;
			i++;
		}
		else{
			arr[k] = b[j];
			k++;
			j++;
		}
	}
	while(i < n1){
		arr[k] = a[i];
		k++;
		i++;
	}
	while(j < n2){
		arr[k] = b[j];
		k++;
		j++;
	}
}
void mergeSort(int arr[], int l , int r){//This function will break the array into subparts until it can't be broken further.
	
	if(l < r){//break it until the left and right side cross each other.
		
		int mid = (l + r)/2;
		mergeSort(arr, l , mid);
		mergeSort(arr, mid + 1, r);
		
		merge(arr, l , mid, r);
	}
}
int main(){
	int arr[] = {5,4,3,2,1};
	mergeSort(arr, 0 ,4);
	for(int i = 0; i < 5; i++){
		cout<<arr[i]<<endl;
	}
	return 0;
}

File: Java/Divide and conquer/merge_sort.java
public class merge_sort {
    // public static void mergeSort(int nums[], int start, int end) {
    // if (start < end) {
    // int mid = start + (end - start) / 2;
    // mergeSort(nums, start, mid);
    // mergeSort(nums, mid + 1, end);
    // merge(nums, start, mid, end);
    // }
    // }

    // public static void merge(int arr[], int start, int mid, int end) {

    // int n1 = mid - start + 1;
    // int n2 = end - mid;
    // int arr1[] = new int[n1];
    // int arr2[] = new int[n2];

    // for (int i = 0; i < n1; i++) {
    // arr1[i] = arr[start + i];
    // }

    // for (int j = 0; j < n2; j++) {
    // arr2[j] = arr[mid + 1 + j];
    // }

    // int i = 0;
    // int j = 0;
    // int k = start;
    // while (i < n1 && j < n2) {
    // if (arr1[i] <= arr2[j]) {
    // arr[k] = arr1[i];
    // i++;
    // } else {
    // arr[k] = arr2[j];
    // j++;
    // }
    // k++;
    // }

    // while (i < n1) {
    // arr[k] = arr1[i];
    // i++;
    // k++;
    // }

    // while (j < n2) {
    // arr[k] = arr2[j];
    // j++;
    // k++;
    // }
    // }

    public static void mergeSort(int nums[], int start, int end) {
        if (start >= end) {
            return;
        }

        int mid = start + (end - start) / 2;
        mergeSort(nums, start, mid);
        mergeSort(nums, start + 1, end);
        merge(nums, start, mid, end);
    }

    public static void merge(int nums[], int start, int mid, int end) {
        int temp[] = new int[end - start + 1];
        int i = start;
        int j = mid + 1;
        int k = 0;

        while (i <= mid && j < end) {
            if (nums[k] < nums[j]) {
                temp[k] = nums[i];
                i++;
            } else {
                temp[k] = nums[j];
                j++;
            }
            k++;
        }

        while (i <= mid) {
            temp[k++] = nums[i++];
        }
        while (j <= end) {
            temp[k++] = nums[j++];
        }

        for (k = 0, i = start; k < temp.length; k++, i++) {
            nums[i] = temp[k];
        }
    }

    public static void main(String[] args) {
        int nums[] = { 6, 3, 9, 2, 5, 8 };
        mergeSort(nums, 0, nums.length - 1);

        for (int i = 0; i < nums.length; i++) {
            System.out.println(nums[i]);
        }
    }
}
